integration-tests/test/ovm-self-upgrades.spec.ts

```diff
@@ -0,0 +1,202 @@
+      await Factory__ReturnTwo.deploy()
+    ).deployTransaction.wait()
+    DeployedBytecode__ReturnTwo = await l2Provider.getCode(
+      returnTwo.contractAddress
+    )
+
+    Factory__SimpleStorage = await ethers.getContractFactory(
+      'SimpleStorage',
+      l2Wallet
+    )
+  })
+
+  describe('setStorage and setCode are correctly applied according to geth RPC', () => {
+    it('Should execute a basic storage upgrade', async () => {
+      const basicStorageUpgrade: ChugSplashInstructions = [
+        {
+          target: OVM_ETH_ADDRESS,
+          key:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+          value:
+            '0x6789123412341234123412341234123412341234123412341234678967896789',
+        },
+      ]
+      await applyAndVerifyUpgrade(basicStorageUpgrade)
+    })
+
+    it('Should execute a basic upgrade overwriting existing deployed code', async () => {
+      const DummyContract = await (
+        await ethers.getContractFactory('SimpleStorage', l2Wallet)
+      ).deploy()
+      await DummyContract.deployTransaction.wait()
+
+      const basicCodeUpgrade: ChugSplashInstructions = [
+        {
+          target: DummyContract.address,
+          code:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+        },
+      ]
+      await applyAndVerifyUpgrade(basicCodeUpgrade)
+    })
+
+    it('Should execute a basic code upgrade which is not overwriting an existing account', async () => {
+      // TODO: fix me?  Currently breaks due to nil pointer dereference; triggerd by evm.StateDB.SetCode(...) in ovm_state_manager.go ?
+      // More recent update: I cannot get this to error out any more.
+      const emptyAccountCodeUpgrade: ChugSplashInstructions = [
+        {
+          target: '0x5678657856785678567856785678567856785678',
+          code:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+        },
+      ]
+      await applyAndVerifyUpgrade(emptyAccountCodeUpgrade)
+    })
+  })
+
+  describe('Contracts upgraded with setStorage and setCode behave as expected', () => {
+    it('code with updated storage returns the new storage', async () => {
+      const SimpleStorage: Contract = await Factory__SimpleStorage.deploy()
+      await SimpleStorage.deployTransaction.wait()
+
+      const valueBefore = await SimpleStorage.value()
+      expect(valueBefore).to.eq(ethers.constants.HashZero)
+
+      const newValue = '0x' + '00'.repeat(31) + '01'
+      const storageVarUpgrade: ChugSplashInstructions = [
+        {
+          target: SimpleStorage.address,
+          key: ethers.constants.HashZero,
+          value: newValue,
+        },
+      ]
+
+      await applyAndVerifyUpgrade(storageVarUpgrade)
+
+      const valueAfter = await SimpleStorage.value()
+      expect(valueAfter).to.eq(newValue)
+    })
+
+    it('code with an updated constant returns the new constant', async () => {
```

Comment:
I'd like to see a few more aggressive tests that change mission-critical code, e.g., change `OVM_ExecutionManager` itself.

Reply:
Here's a list I can think of:
1. Modify `OVM_ExecutionManager`.
2. Modify storage in `OVM_ExecutionManager`.
3. Modify the `OVM_SequencerEntrypoint`, `OVM_ProxyEOA`, and `OVM_ECDSAContractAccount`.
4. Modify `OVM_ETH`.

Reply:
WRT 2. here: I actually think that modifying storage of the EM is actually something that is off-limits.  Because it isn't virtualized within the `OVM_StateManager.contractStorage` mapping, it's gonna be a pain to get that into L1 correctly.  I think that the correct model (also relevant for L1 refunds if storing `0x00...` and minimizing cost during the fraud proof) we should follow is that EM storage slots are always reset (i.e. `EM._resetContext`) at the end of execution.  This also means we should consider setting the first 50 slots to a nonzero value in case we want more variables in the future.